### PR TITLE
Allow multiple named shared memory regions

### PR DESCRIPTION
**Pull request by TheMarex**
# Issue

This PR adds support for having multiple named shared memory regions.
This is going to unlock https://github.com/Project-OSRM/osrm-backend/issues/1221 and https://github.com/Project-OSRM/osrm-backend/issues/4873 by making it easier to have multiple named regions.

## Tasklist

 - [x] CHANGELOG.md entry ([How to write a changelog entry](http://keepachangelog.com/en/1.0.0/#how))
 - [x] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [x] add tests (see [testing documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/testing.md)
 - [x] review
 - [x] adjust for comments

## Requirements / Relations

Based on #4975.

### CODE REVIEW

File: include/storage/shared_datatype.hpp [hunk 1]
```diff
@@ -160,40 +160,111 @@ class DataLayout
...
+        for (const auto &region : regions)
+        {
+            if (!region.IsEmpty())
+            {
+                *out++ = region.name;
+            }
+        }
+    }
+
+    void Deregister(const RegionID key) { regions[key] = SharedRegion{}; }
```
Comment (oxidase) [inline include/storage/shared_datatype.hpp]: The method is not used and `DataWatchdogImpl` assumes that the shared region pointer won't be changed during an engine process life-time. Is a follow-up issue about deleting named regions needed?

File: src/osrm/osrm.cpp [hunk 2]
```diff
@@ -28,12 +28,6 @@ OSRM::OSRM(engine::EngineConfig &config)
         throw util::exception("Required files are missing, cannot continue.  Have all the "
                               "pre-processing steps been run?");
     }
-    else if (config.use_shared_memory)
```
Comment (oxidase) [inline src/osrm/osrm.cpp]: :+1:

File: include/storage/shared_datatype.hpp (hunk 1 above)
Comment (TheMarex) [reply include/storage/shared_datatype.hpp]: Good catch, I wanted to use this for the spring-clean command, but instead we just overwrite the whole registry.

**Comment by oxidase** [comment]
It is also possible to add `features/options/datastore/datastore.feature` options tests to increase coverage 
```
@datastore @options @help
Feature: osrm-datastore command line options: list

    Background:
        Given the profile "testbot"
        And the node map
            """
            a b
            """
        And the ways
            | nodes |
            | ab    |
        And the data has been contracted

    Scenario: osrm-datastore - Help should be shown when no options are passed
        When I try to run "osrm-datastore --dataset-name test_dataset_42 {processed_file}"
        Then it should exit successfully
        When I try to run "osrm-datastore --list"
        Then it should exit successfully
        And stdout should contain "test_dataset_42/data"
```

**Comment by TheMarex** [comment]
@oxidase thanks for the test case! Changed all locations and added the test. Should be good to merge after Travis passes.

**Pull request merged**